flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGFailoverTest.java

```diff
@@ -0,0 +1,763 @@
+	public void testRegionFailoverInEagerMode() throws Exception {
+
+		// create a streaming job graph with EAGER schedule mode
+		final JobGraph jobGraph = createStreamingJobGraph();
+		final long checkpointId = 42L;
+		final ExecutionGraph eg = createExecutionGraph(jobGraph, checkpointId);
+
+		final Iterator<ExecutionVertex> vertexIterator = eg.getAllExecutionVertices().iterator();
+		final ExecutionVertex ev11 = vertexIterator.next();
+		final ExecutionVertex ev12 = vertexIterator.next();
+		final ExecutionVertex ev21 = vertexIterator.next();
+		final ExecutionVertex ev22 = vertexIterator.next();
+
+		// verify initial state
+		assertVertexInState(ExecutionState.DEPLOYING, ev11);
+		assertVertexInState(ExecutionState.DEPLOYING, ev12);
+		assertVertexInState(ExecutionState.DEPLOYING, ev21);
+		assertVertexInState(ExecutionState.DEPLOYING, ev22);
+
+		// verify initial attempt number
+		assertEquals(0, ev11.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(0, ev12.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(0, ev21.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(0, ev22.getCurrentExecutionAttempt().getAttemptNumber());
+
+		// trigger task failure of ev11
+		// vertices { ev11, ev21, ev22 } should be affected
+		testMainThreadUtil.execute(() -> ev11.getCurrentExecutionAttempt().fail(new Exception("Test Exception")));
+
+		// verify vertex states and complete cancellation
+		assertVertexInState(ExecutionState.FAILED, ev11);
+		assertVertexInState(ExecutionState.DEPLOYING, ev12);
+		assertVertexInState(ExecutionState.CANCELING, ev21);
+		assertVertexInState(ExecutionState.DEPLOYING, ev22);
+		testMainThreadUtil.execute(() -> ev21.getCurrentExecutionAttempt().completeCancelling());
+
+		// verify vertex states
+		// in eager mode, all affected vertices should be scheduled in failover
+		assertVertexInState(ExecutionState.DEPLOYING, ev11);
+		assertVertexInState(ExecutionState.DEPLOYING, ev12);
+		assertVertexInState(ExecutionState.DEPLOYING, ev21);
+		assertVertexInState(ExecutionState.DEPLOYING, ev22);
+
+		// verify attempt number
+		assertEquals(1, ev11.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(0, ev12.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(1, ev21.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(0, ev22.getCurrentExecutionAttempt().getAttemptNumber());
+	}
+
+	/**
+	 * Tests for scenes that a task fails for its own error, in which case the
+	 * region containing the failed task and its consumer regions should be restarted.
+	 * <pre>
+	 *     (v11) --> (v21) -+-> (v31) --> (v41)
+	 *                      x
+	 *     (v12) --> (v22) -+-> (v32) --> (v42)
+	 *
+	 *                      ^
+	 *                      |
+	 *                  (blocking)
+	 * </pre>
+	 * 4 regions. Each region has 2 pipelined connected vertices.
+	 */
+	@Test
+	public void testRegionFailoverForRegionInternalErrorsInLazyMode() throws Exception {
```

Comment:
Compared to `testRegionFailoverInEagerMode()`, what does this test add? As far as I can see, additional lines that are covered are:

```
case LAZY_FROM_SOURCES:
	newSchedulingFuture = AdaptedSchedulingUtils.scheduleLazy(vertices, executionGraph);
	break;
```

Is it justified to complicate the testing job topology for these additional lines, or can these be tested differently?

Reply:
Will simplify the topology to 4 regions with one vertex each.
There were 2 purposes to complicate the topology:
1. cover the case that each region has more than one vertices. It's not necessary anyhow
2. make it easier to control the cancelling termination. Yet we have other approaches to do this